### PR TITLE
[ROCm] LoadHIP CMake cleanup

### PR DESCRIPTION
**Pull request by naromero77amd**
Should help mitigate issues reported here: https://github.com/pytorch/pytorch/issues/128313

While working on https://github.com/pytorch/pytorch/pull/136700, we realized that some of the ROCm CMake can be streamlined. 

This PR does not fix any bugs or provide any new functionality. Strictly clean-up.

The remaining `${ROCM_ROCTX_LIB}` will be removed when we transition to the rocprofiler-sdk (to be done in a separate PR).

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Comment by naromero77amd** [comment]
FWIW tested against ROCm 6.1 and still builds successfully.

**Review by jithunnair-amd — APPROVED** [review]
LGTM, thanks for the cleanup!

**Comment by naromero77amd** [comment]
@jithunnair-amd 

I would propose getting rid of these two sets of lines as well:

We are *already* getting the this same information later from all the `find_package_and_print_version`.
https://github.com/pytorch/pytorch/blob/419baf47ac2f922117d11d5339961f77a39f92ac/cmake/public/LoadHIP.cmake#L53-L122

I am not sure what was the original purpose of these line, but they also provide duplicate information except for the one for `hsakmt-roct` (since don't we don't do a `find_package_and_print_version(hsakmt-roct)`. `dpkg` is specific to Debian/Ubuntu and they would not work on other Linux distributions. 
https://github.com/pytorch/pytorch/blob/419baf47ac2f922117d11d5339961f77a39f92ac/cmake/public/LoadHIP.cmake#L124-L131

Let me know if you agree that we should get rid of both groups of CMake lines.

**Comment by jithunnair-amd** [comment]
> Let me know if you agree that we should get rid of both groups of CMake lines.

**Group 1:**
For the first group, it serves two purposes:
1. Display ROCM_VERSION by detecting rocm_version.h: This part can surely be cleaned up by a) removing references to old rocm_version.h file location, b) removing "caffe2" in the messages and c) possibly using alternate cmake file read/grep instead of the current approach of creating a `detect_rocm_version.cc` file solely to extract `ROCM_VERSION_[MAJOR|MINOR|PATCH]` from rocm_version.h into CMake
2. Set [ROCM_VERSION](https://github.com/pytorch/pytorch/blob/711dacf9845cbc9ea8b3b0fa257309930106712f/cmake/Dependencies.cmake#L1069) and [TORCH_HIP_VERSION](https://github.com/pytorch/pytorch/blob/711dacf9845cbc9ea8b3b0fa257309930106712f/cmake/Dependencies.cmake#L1070) for the CMake build flow. These are heavily used in PyTorch sources to conditionalize code based on ROCm version and wherever CUDA_VERSION is hipified to TORCH_HIP_VERSION (to enable latest CUDA version code for ROCm by default, since CUDA_VERSION is in range of 11xxx or more, while TORCH_HIP_VERSION is in range of 6xx or more)

For these reasons, we cannot remove but can refactor the first group.

**Group 2:**
I agree that using `dpkg` is OS-restrictive, however those OS package versions serve more as a debug aid to tell us more about the ROCm installation in the user environment. We can surely remove the `hcc`, `hip-base` and `hip_hcc` lines, since those are obsolete, but the remaining OS package versions don't directly map to the ROCm library versions printed in the next section.

**Comment by naromero77amd** [comment]
@jithunnair-amd Please re-review. Thanks.

File: cmake/public/LoadHIP.cmake [hunk 1]
```diff
@@ -44,78 +44,64 @@ endif()
...
   endif()
 
+  # Read the ROCM headerfile into a variable
+  file(READ ${ROCM_HEADER_FILE} ROCM_HEADER_CONTENT)
+
+  # Below we use a RegEx to find ROCM version numbers.
+  # Note that CMake does not support \s for blank space. That is
+  # why in the regular expressions below we have a blank space in
+  # the square brackets.
+  # There are three streps:
```
Comment (jithunnair-amd) [inline cmake/public/LoadHIP.cmake:85]: typo "steps"

File: cmake/public/LoadHIP.cmake [hunk 2]
```diff
@@ -44,78 +44,64 @@ endif()
...
+  # 1. Match regular expression
+  # 2. Strip thie non-numerical part of the string
```
Comment (jithunnair-amd) [inline cmake/public/LoadHIP.cmake:87]: typo "this" or "the"?

File: cmake/public/LoadHIP.cmake [hunk 3]
```diff
@@ -44,78 +44,64 @@ endif()
...
+  string(STRIP ${TEMP2} ROCM_VERSION_DEV_MINOR)
+  string(REGEX MATCH "ROCM_VERSION_PATCH [ ]*[0-9]+" TEMP1 ${ROCM_HEADER_CONTENT})
+  string(REPLACE "ROCM_VERSION_PATCH" "" TEMP2 ${TEMP1})
+  string(STRIP ${TEMP2} ROCM_VERSION_DEV_PATCH)
+
+  # Create ROCM_VERSION_DEV_INT which is later used as a preprocessor macros
+  set(ROCM_VERSION_DEV "${ROCM_VERSION_DEV_MAJOR}.${ROCM_VERSION_DEV_MINOR}.${ROCM_VERSION_DEV_PATCH}")
+  math(EXPR ROCM_VERSION_DEV_INT "(${ROCM_VERSION_DEV_MAJOR}*10000) + (${ROCM_VERSION_DEV_MINOR}*100) + ${ROCM_VERSION_DEV_PATCH}")
+
+  message(STATUS "Caffe2: Header version is: " ${ROCM_VERSION_DEV})
```
Comment (jithunnair-amd) [inline cmake/public/LoadHIP.cmake:103]: Caffe2 is obsolete anyway, plus not really sure why it should say Caffe2 in there. Let's make this "ROCm Header version is: "

**Review by jithunnair-amd — APPROVED** [review]
Please fix typos and nits. Overall, the parsing of ROCm version reads better and is less convoluted IMO!

**Review by jeffdaily — APPROVED** [review]

**Pull request closed without merging**